### PR TITLE
[RFC] Fix PVS analysis warnings

### PR DESCRIPTION
**Pull request by ZyX-I**
Left only four of them:

1. Mysterious “unreachable code” in msgpack_rpc_from_object (line 389).
2. Buffer undeflow in spell.c:3638: not sure what to do with this exactly (may leave as-is for sanitizers to catch some possible bugs or just zero out the entire stack).
3. Parameter rewrite in syntax.c:2186: almost sure it is intentional, but I am not familiar with this part of code.
4. Potential NULL dereference at regexp.c:6932. There is no NULL dereference there as far as I see, but what there is is a two-interation for loop antipattern (for loop: compute length then do allocation and fill the buffer).

### CODE REVIEW

File: test/functional/fixtures/tty-test.c [hunk 1]
```diff
@@ -6,6 +6,9 @@
 #include <stdlib.h>
 #include <uv.h>
 
+// -V:STRUCT_CAST:641
+#define STRUCT_CAST(Type, obj) ((Type *)(obj))
```
Comment (justinmk) [inline test/functional/fixtures/tty-test.c]: Does this hide strict aliasing violations?

Comment (ZyX-I) [inline test/functional/fixtures/tty-test.c]: This does not hide anything, just silences some warnings. If I am not mistaking regarding what strict aliasing means then you may search for STRUCT_CAST to find code which for sure violates strict aliasing rules. It is libuv which makes us play with struct pointers though.

**Comment by ZyX-I** [comment]
Note regarding commit messages: “silencing” means “I think there is a false positive” and “fixing” means that I think analyzer found a real bug, or, at least, stinky code.

File: src/nvim/charset.c [hunk 2]
```diff
@@ -545,18 +545,8 @@ void transchar_nonprint(char_u *buf, int c)
     buf[1] = (char_u)(c ^ 0x40);
 
     buf[2] = NUL;
-  } else if (c >= 0x80) {
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Based on the Vim code it appears that `|x` and `~x` were intended for non-multibyte builds or when `&encoding` is not UTF-8: not actually a dead code, just we removed all conditions which may trigger it.

File: test/functional/fixtures/tty-test.c (hunk 1 above)
Comment (justinmk) [inline test/functional/fixtures/tty-test.c]: Yeah, but I wonder if it could hide something in the future. Maybe we should leave the PVS warnings.

Comment (ZyX-I) [inline test/functional/fixtures/tty-test.c]: How? For the compiler it expands to the same thing it was before introducing the macros plus two pairs of parenthesis. Do not think this should be a problem. Also PVS warnings are false positives and AFAIR we agreed that goal should be zero warnings on two highest certainty levels.

**Comment by ZyX-I** [comment]
CI have passed.

File: test/functional/fixtures/tty-test.c (hunk 1 above)
Comment (justinmk) [inline test/functional/fixtures/tty-test.c]: > How?

I mean if, e.g., the type argument to the parameter is changed for some reason in the future.

**Pull request merged**